### ISSUE TITLE
[Development Call]: 

### ISSUE THREAD

**Issue by NielsRahder** [tilburgsciencehub/website#798]
### Contact Details

N.J.Rahder@tilburguniversity.edu

### Is your content request related to a problem you've encountered during your research process? Please describe.

To construct a more comprehensive framework for machine learning, it's important to establish a series of incremental building blocks. These incremental blocks (stepping stones) serve as a gradual and intuitive path for beginners, rather than plunging them directly into the deep end. The additional building blocks I propose (with different sub-chapters) are the following.

Decision trees (ranger package in R?)
- basic concept 
- how a tree is built (splitting with (e.g. Gini impurity/ information gain)
- tree pruning
- handling numerical v/s categorical features
- visualization 

Model evaluation 
- basic concept 
- different evaluation methods (e.g., accuracy, precision, ROC/AUC)
- cross-validation 

Hyperparameter tuning
- basic concept 
- different tuning methods (e.g., grid search, random search, Bayesian)

Ensemble methods
- basic concept 
- elaborate on random forests
- bagging (simultaneous)
- boosting (sequential)

If anyone has any comments or suggestions I would love to hear! 

### Would you like this to be a Building Block or a Tutorial?

None

### Suggested title

Create new building blocks for machine learning 

### Suggested keywords

Machine learning, stepping stones 

### Required tools

R (and potentially Python)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ana-bianca-luca** [comment]
This is a large topic and should be organized properly - perhaps start with a large overview of machine learning (and how it is different from deep learning), an introduction to supervised and unsupervised learning (describe each), what models each contains (regression/decision trees/random forest/kNN etc).
Perhaps a good approach is to create two chapters in a machine learning tab, one for supervised and one for unsupervised learning, and under each chapter to have separate bb's for each models they contain

**Comment by NielsRahder** [comment]
Thanks, @ana-bianca-luca for the remarks!

I really like your idea about the separate chapters, and will definitely look into it. 
About the Python remark: I started with R as I have more experience with that, but it was still on my to-do to add the Python code.

**Comment by Fernando-Iscar** [comment]
Following up on the conversation in the last Wow meeting, and the comments on this issue, here are some potential ideas about building blocks and tutorials related to this topic.

First of all, currently Machine Learning material is just one introductory BB which only contains a link to another website tutorial. We can get some inspiration from that website, but in my opinion, and following @ana-bianca-luca's suggestion, we should re-structure it in another tab called Machine Learning.

# Building Blocks

This section could have at least two subsections, under each some BB topic ideas:

## Supervised Learning

- **Introduction to Supervised Learning** 
    -  Introduction to supervised learning concepts and techniques
- **Regression:** 
    - Theory, linear regression example, brief mention of other regressions like lasso, ridge, etc.
- **Classification:**
    - Theory, KNN example, brief mention of other algorithms like DT, SVM, Logistic Regression, Naive-Bayes, etc.
- **Model Evaluation:** 
    - Regression metrics (R2, MSE, MAE, etc.)
    - Classification metrics (accuracy, precision, recall, F1-score, ROC-AUC.)

## Unsupervised Learning

- **Intro to UL to Unsupervised Learning**
    -  Introduction to unsupervised learning concepts and techniques
- **Clustering Basics**
   - K-Means, Hierarchical and Spectral Clustering...
- **Dimensionality Reduction**
   - Theory, PCA example, brief mention of other techniques like LLE, t-SNE, etc.
- **Association**
   - Understanding concepts like apriori algorithm, FP-growth, etc. with examples demonstrating market basket analysis.

# Tutorials

Since it is a broad topic, for extended content we could make some tutorials like:

- **Build a classification model for Fraud Detection:**

    - A step-by-step guide on how to build, train, fine-tune and evaluate a classification model aimed at detecting fraud in a dataset, considering different features and data imbalances.

- **Dimensionality Reduction in Data Visualization:**

    - Illustrating how to utilize dimensionality reduction techniques like PCA and t-SNE for data visualization, facilitating the understanding of complex, high-dimensional data through graphical representations

We could also think of including Deep Learning ideas or a separate section, BB on hyperparameter tuning, etc. But for now, this could be a starting point.

To make sure everyone can easily understand and use these concepts, it would be good to have material available for both R and Python users. We could think about covering each topic twice, once for each programming language, to facilitate a better grasp for users of both languages.


Looking forward to hearing your thoughts

@NielsRahder 
@hannesdatta

**Comment by NielsRahder** [comment]
@Fernando-Iscar the division and suggestions look good to me as a starting point. It lets us easily add more content later on.  

With regard to making the material available for both R and Python I would suggest using the "dual code blocks" (see BB  ["Download Data Programmatically"](https://tilburgsciencehub.com/building-blocks/store-and-document-your-data/store-data/download-data/) ") so that we can keep our topics more organized (Instead of having to create multiple BB's / Tutorials per topic).
